src/test/denialofservice_tests.cpp

```diff
@@ -391,7 +419,14 @@ BOOST_AUTO_TEST_CASE(DoS_mapOrphans)
     // ... and 50 that depend on other orphans:
     for (int i = 0; i < 50; i++)
     {
-        CTransactionRef txPrev = RandomOrphan();
+        CTransactionRef txPrev;
+        if (i == 0) {
+            // This block makes sure that the condition "if (it == mapOrphanTransactions.end())" in OrphanByIndex() gets called at least once.
+            // Otherwise test coverage is non-deterministic.
+            txPrev = OrphanByIndex(ArithToUint256(arith_uint256(std::numeric_limits<uint64_t>::max())));
```

Comment:
instead of hardcoding special cases, I'd prefer if the txids were just derived from a deterministic fast random context. If you want, you can explicitly seed it so that the condition is hit at least once.

Reply:
Ok, I can do that. The RNG can indeed be seeded so that the condition is hit at least once. However, the `ForceCoverageInPubKeyVerify` function must remain, since the non-determinism is not caused by different seeds, but is there due to concurrency issues (please refer to my comment [here](https://github.com/bitcoin/bitcoin/pull/16878?_pjax=%23js-repo-pjax-container#discussion_r331741093) for some details). I therefore don't see another way to remove the non-determinism other than this. Is that ok?

Reply:
Generally the tests run in a single thread, so there shouldn't be any concurrency issues. Are you sure the non-determinism is not caused by the random private key, which can also be hardcoded?

Reply:
Sorry for the long delay, you are right, the additional non-determinism was caused by the fact that the private key is generated by another RNG and not by concurrency issues. I removed the function I added to force coverage and forced the key to be generated by the same RNG for this test. I seeded it so that all branches of the function `ecdsa_signature_parse_der_lax` run. However, after a few dozen times of running this test with the script `contrib/devtools/test_deterministic_coverage.sh`, there was another place where the coverage was not deterministic, in the function `Loop()` in `checkqueue.h`. I'm looking into it.